pkg/queryfrontend/cache/memcached.go

```diff
@@ -0,0 +1,145 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package cache
+
+import (
+	"context"
+	"time"
+
+	cortexcache "github.com/cortexproject/cortex/pkg/chunk/cache"
+	"github.com/cortexproject/cortex/pkg/querier/queryrange"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"gopkg.in/yaml.v2"
+
+	"github.com/thanos-io/thanos/pkg/cacheutil"
+)
+
+const (
+	memcachedDefaultTTL = 24 * time.Hour
+)
+
+// MemcachedCache is a memcached-based cache.
+type MemcachedCache struct {
+	logger    log.Logger
+	memcached cacheutil.MemcachedClient
+	validity  time.Duration
+
+	// Metrics.
+	requests prometheus.Counter
+	hits     prometheus.Counter
+}
+
+type MemcachedResponseCacheConfig struct {
+	Memcached cacheutil.MemcachedClientConfig `yaml:",inline"`
+	Validity  time.Duration                   `yaml:"validity"`
+}
+
+// newMemcachedCache makes a new MemcachedCache and returns Cortex ResultsCacheConfig.
+func newMemcachedCache(conf []byte, logger log.Logger, reg prometheus.Registerer) (*queryrange.ResultsCacheConfig, error) {
+	var config MemcachedResponseCacheConfig
+	if err := yaml.UnmarshalStrict(conf, &config); err != nil {
+		return nil, err
+	}
+
+	memcachedClientConfig, err := yaml.Marshal(config.Memcached)
+	if err != nil {
+		return nil, errors.Wrap(err, "marshal content of memcached client configuration")
+	}
+
+	memcached, err := cacheutil.NewMemcachedClient(logger, "response-cache", memcachedClientConfig, reg)
+	if err != nil {
+		return nil, errors.Wrap(err, "create memcached client")
+	}
+
+	c := newMemcachedCacheWithClient(memcached, config.Validity, logger, reg)
+
+	return &queryrange.ResultsCacheConfig{
+		CacheConfig: cortexcache.Config{Cache: c},
+	}, nil
+}
+
+func newMemcachedCacheWithClient(
+	memcached cacheutil.MemcachedClient,
+	validity time.Duration,
+	logger log.Logger,
+	reg prometheus.Registerer,
+) *MemcachedCache {
+	c := &MemcachedCache{
+		logger:    logger,
+		memcached: memcached,
+		validity:  validity,
+	}
+
+	if c.validity == 0 {
+		level.Info(logger).Log("msg", "memcached cache valid time set to 0, use 24 hours instead")
+		c.validity = memcachedDefaultTTL
+	}
+
+	c.requests = promauto.With(reg).NewCounter(prometheus.CounterOpts{
```

Comment:
We have the same metrics for store cache https://github.com/thanos-io/thanos/blob/798f40b5e7f7bffb5c9713ad72bc70ce2af7e411/pkg/cache/memcached.go#L35

This won't create a registration issue, however better to add a constant label (e.g name) to create distinction between them.

Reply:
Makes sense. If I add the const label, it is necessary to also add the `name` label to the store cache metrics as well?

If it is this case, then maybe let me rename the two metrics to `thanos_response_cache_memcached_requests_total` and `thanos_response_cache_memcached_hits_total`

Reply:
Store cache metrics already have this I believe. You just need to add the name here as well. 

> If it is this case, then maybe let me rename the two metrics to thanos_response_cache_memcached_requests_total and thanos_response_cache_memcached_hits_total

Let's go with the name approach.

Reply:
Thank you, let me update my pr.